Providers/Profile.swift

```diff
@@ -576,28 +592,48 @@ public class BrowserProfile: Profile {
         }
 
         /**
-         * Locking is managed by withSyncInputs. Make sure you take and release these
+         * Locking is managed by syncSeveral. Make sure you take and release these
          * whenever you do anything Sync-ey.
          */
-        var syncLock = OSSpinLock() {
-            didSet {
-                let notification = syncLock == 0 ? NotificationProfileDidFinishSyncing : NotificationProfileDidStartSyncing
-                NSNotificationCenter.defaultCenter().postNotification(NSNotification(name: notification, object: nil))
-            }
-        }
+        private let syncLock = NSRecursiveLock()
 
-        // According to the OSAtomic header documentation, the convention for an unlocked lock is a zero value
-        // and a locked lock is a non-zero value
         var isSyncing: Bool {
-            return syncLock != 0
+            syncLock.lock()
+            defer { syncLock.unlock() }
+            return !(syncTerminal?.isFilled ?? true)
```

Comment:
We nil-out `syncTerminal` within the critical section, immediately after filling it. So we can only ever get here if it's nil or unfilled, right?